`fabric.setup_module` fails on MPS

### Bug description

I am following [this Lightning blogpost](https://lightning.ai/blog/8-bit-quantization-with-lightning-fabric/) to post-training quantize an HuggingFace's model using Lightning Fabric, however the `setup_module` command fails because of this failed assertion: `AssertionError: Torch not compiled with CUDA enabled`. 
I don't have a NVidia GPU on my machine, however the MPS backend, which I am using, should be supported. In fact, I specify in the `Fabric` object that the accelerator to use is MPS. 

Is there a way to make this work without CUDA?

### What version are you seeing the problem on?

v2.1

### How to reproduce the bug

```python
from lightning.fabric import Fabric
from lightning.fabric.plugins import BitsandbytesPrecision

model = YourModel()
plugin = BitsandbytesPrecision(mode="int8")
fabric = Fabric(plugins=plugin, accelerator="mps")
model = fabric.setup_module(model)
```


### Error messages and logs

```
AssertionError: Torch not compiled with CUDA enabled
```


### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component: Fabric, BitsandbytesPrecision
#- PyTorch Lightning Version: v2.1.3
#- PyTorch Version: v2.1.2
#- Python version: v3.11.7
#- OS: MacOS 14.1
#- GPU models and configuration: MPS
#- How you installed Lightning: conda
#- Running environment of LightningApp: local
```

</details>


### More info

_No response_

cc @carmocca @justusschock @awaelchli

--- a/src/lightning/fabric/connector.py
+++ b/src/lightning/fabric/connector.py
@@ -449,7 +449,9 @@ class _Connector:
 
     def _check_and_init_precision(self) -> Precision:
         if isinstance(self._precision_instance, Precision):
-            if isinstance(self._precision_instance, BitsandbytesPrecision) and not isinstance(self.accelerator, CUDAAccelerator):
+            if isinstance(self._precision_instance, BitsandbytesPrecision) and not isinstance(
+                self.accelerator, CUDAAccelerator
+            ):
                 raise RuntimeError("Bitsandbytes is only supported on CUDA GPUs.")
             return self._precision_instance
         if isinstance(self.strategy, (SingleDeviceXLAStrategy, XLAStrategy, XLAFSDPStrategy)):

--- a/src/lightning/fabric/connector.py
+++ b/src/lightning/fabric/connector.py
@@ -24,6 +24,7 @@ from lightning.fabric.accelerators.cuda import CUDAAccelerator
 from lightning.fabric.accelerators.mps import MPSAccelerator
 from lightning.fabric.accelerators.xla import XLAAccelerator
 from lightning.fabric.plugins import (
+    BitsandbytesPrecision,
     CheckpointIO,
     DeepSpeedPrecision,
     HalfPrecision,
@@ -448,6 +449,8 @@ class _Connector:
 
     def _check_and_init_precision(self) -> Precision:
         if isinstance(self._precision_instance, Precision):
+            if isinstance(self._precision_instance, BitsandbytesPrecision) and not isinstance(self.accelerator, CUDAAccelerator):
+                raise RuntimeError("Bitsandbytes is only supported on CUDA GPUs.")
             return self._precision_instance
         if isinstance(self.strategy, (SingleDeviceXLAStrategy, XLAStrategy, XLAFSDPStrategy)):
             return XLAPrecision(self._precision_input)  # type: ignore

--- a/src/lightning/pytorch/trainer/connectors/accelerator_connector.py
+++ b/src/lightning/pytorch/trainer/connectors/accelerator_connector.py
@@ -566,7 +566,9 @@ class _AcceleratorConnector:
 
     def _validate_precision_choice(self) -> None:
         """Validate the combination of choices for precision, AMP type, and accelerator."""
-        if isinstance(self._precision_plugin_flag, BitsandbytesPrecision) and not isinstance(self.accelerator, CUDAAccelerator):
+        if isinstance(self._precision_plugin_flag, BitsandbytesPrecision) and not isinstance(
+            self.accelerator, CUDAAccelerator
+        ):
             raise RuntimeError("Bitsandbytes is only supported on CUDA GPUs.")
         if _habana_available_and_importable():
             from lightning_habana import HPUAccelerator

--- a/src/lightning/pytorch/trainer/connectors/accelerator_connector.py
+++ b/src/lightning/pytorch/trainer/connectors/accelerator_connector.py
@@ -37,6 +37,7 @@ from lightning.pytorch.accelerators.mps import MPSAccelerator
 from lightning.pytorch.accelerators.xla import XLAAccelerator
 from lightning.pytorch.plugins import (
     _PLUGIN_INPUT,
+    BitsandbytesPrecision,
     CheckpointIO,
     DeepSpeedPrecision,
     DoublePrecision,
@@ -565,6 +566,8 @@ class _AcceleratorConnector:
 
     def _validate_precision_choice(self) -> None:
         """Validate the combination of choices for precision, AMP type, and accelerator."""
+        if isinstance(self._precision_plugin_flag, BitsandbytesPrecision) and not isinstance(self.accelerator, CUDAAccelerator):
+            raise RuntimeError("Bitsandbytes is only supported on CUDA GPUs.")
         if _habana_available_and_importable():
             from lightning_habana import HPUAccelerator
 

--- a/tests/tests_fabric/test_connector.py
+++ b/tests/tests_fabric/test_connector.py
@@ -13,6 +13,7 @@
 # limitations under the License
 import inspect
 import os
+import sys
 from typing import Any, Dict
 from unittest import mock
 from unittest.mock import Mock
@@ -29,6 +30,7 @@ from lightning.fabric.accelerators.cuda import CUDAAccelerator
 from lightning.fabric.accelerators.mps import MPSAccelerator
 from lightning.fabric.connector import _Connector
 from lightning.fabric.plugins import (
+    BitsandbytesPrecision,
     DeepSpeedPrecision,
     DoublePrecision,
     FSDPPrecision,
@@ -862,6 +864,13 @@ def test_precision_selection_amp_ddp(strategy, devices, is_custom_plugin, plugin
         plugins=plugin,
     )
     assert isinstance(connector.precision, plugin_cls)
+
+
+def test_bitsandbytes_precision_cuda_required(monkeypatch):
+    monkeypatch.setattr(lightning.fabric.plugins.precision.bitsandbytes, "_BITSANDBYTES_AVAILABLE", True)
+    monkeypatch.setitem(sys.modules, "bitsandbytes", Mock())
+    with pytest.raises(RuntimeError, match="Bitsandbytes is only supported on CUDA GPUs"):
+        _Connector(accelerator="cpu", plugins=BitsandbytesPrecision(mode="int8"))
 
 
 @pytest.mark.parametrize(("strategy", "strategy_cls"), [("DDP", DDPStrategy), ("Ddp", DDPStrategy)])

--- a/tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py
+++ b/tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py
@@ -18,6 +18,7 @@ from typing import Any, Dict
 from unittest import mock
 from unittest.mock import Mock
 
+import lightning.fabric
 import lightning.pytorch
 import pytest
 import torch

--- a/tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py
+++ b/tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py
@@ -35,6 +35,7 @@ from lightning.pytorch.accelerators import Accelerator, CPUAccelerator, CUDAAcce
 from lightning.pytorch.plugins.io import TorchCheckpointIO
 from lightning.pytorch.plugins.layer_sync import LayerSync, TorchSyncBatchNorm
 from lightning.pytorch.plugins.precision import (
+    BitsandbytesPrecision,
     DeepSpeedPrecision,
     DoublePrecision,
     FSDPPrecision,
@@ -1115,3 +1116,10 @@ def test_connector_num_nodes_input_validation():
 def test_precision_selection(precision_str, strategy_str, expected_precision_cls):
     connector = _AcceleratorConnector(precision=precision_str, strategy=strategy_str)
     assert isinstance(connector.precision_plugin, expected_precision_cls)
+
+
+def test_bitsandbytes_precision_cuda_required(monkeypatch):
+    monkeypatch.setattr(lightning.fabric.plugins.precision.bitsandbytes, "_BITSANDBYTES_AVAILABLE", True)
+    monkeypatch.setitem(sys.modules, "bitsandbytes", Mock())
+    with pytest.raises(RuntimeError, match="Bitsandbytes is only supported on CUDA GPUs"):
+        _AcceleratorConnector(accelerator="cpu", plugins=BitsandbytesPrecision(mode="int8"))

